src/main/java/org/sagebionetworks/bridge/services/AdherenceService.java

```diff
@@ -549,4 +559,223 @@ public AdherenceStatistics getAdherenceStatistics(String appId, String studyId,
+            
+            if (scheduledAssessmentMap.containsKey(instanceGuid)) {
+                DetailedAdherenceReportAssessmentRecord assessmentRecord = new DetailedAdherenceReportAssessmentRecord();
+                
+                String refKey = scheduledAssessmentMap.get(instanceGuid).getRefKey();
+                AssessmentInfo assessmentInfo = assessmentInfoMap.get(refKey);
+                AssessmentReference assessmentReference = assessmentReferenceMap.get(assessmentInfo.getGuid());
+                
+                assessmentRecord.setAssessmentInstanceGuid(instanceGuid);
+                assessmentRecord.setAssessmentId(assessmentInfo.getIdentifier());
+                assessmentRecord.setAssessmentGuid(assessmentInfo.getGuid());
+                assessmentRecord.setAssessmentName(assessmentReference.getTitle());
+                
+                if (record.getStartedOn() != null) {
+                    assessmentRecord.setAssessmentStart(record.getStartedOn().withZone(state.getTimeZone()));
+                }
+                if (record.getFinishedOn() != null) {
+                    assessmentRecord.setAssessmentCompleted(record.getFinishedOn().withZone(state.getTimeZone()));
+                }
+                if (record.getUploadedOn() != null) {
+                    assessmentRecord.setAssessmentUploadedOn(record.getUploadedOn().withZone(state.getTimeZone()));
+                }
+                
+                
+                if (record.isDeclined()) {
+                    assessmentRecord.setAssessmentStatus("Declined");
+                } else if (record.getFinishedOn() != null) {
+                    assessmentRecord.setAssessmentStatus("Completed");
+                } else if (record.getStartedOn() != null) {
+                    assessmentRecord.setAssessmentStatus("Not Completed");
+                }
+                
+                assessmentRecord.setSortPriority(sortReferenceMap.get(instanceGuid));
+                
+                String parentInstanceGuid = assessmentParentMap.get(instanceGuid);
+                ScheduledSession scheduledSession = scheduledSessionMap.get(parentInstanceGuid);
+                
+                fillSessionRecord(sessionRecords, scheduledSession, //sessionMap, 
+                        state, sortReferenceMap, null);
+                
+                Map<String, DetailedAdherenceReportAssessmentRecord> assessmentRecords = sessionRecords
+                        .get(scheduledSession.getInstanceGuid()).getAssessmentRecordMap();
+                assessmentRecords.put(assessmentRecord.getAssessmentInstanceGuid(), assessmentRecord);
+                
+            } else if (scheduledSessionMap.containsKey(instanceGuid)) {
+                ScheduledSession scheduledSession = scheduledSessionMap.get(instanceGuid);
+                
+                fillSessionRecord(sessionRecords, scheduledSession, state, sortReferenceMap, record);
+            }
```

Comment:
It looks like this is only possible if we have scheduled sessions that have no session assessments. Is that possible? Should we have a unit test for this scenario?

Reply:
I actually don't know if it's possible to have a session without assessments. I haven't seen that.

This section is comparing the instanceGuid from the adherence record to the timeline. And the adherence record could reference either an assessment instance or a session instance. This block takes the session adherence records and uses them to fill out startedOn/finishedOn information in the session report records.
The tricky part is that the adherence records can come in any order, so I couldn't guarantee that the session record would exist before its child assessments are created. So a placeholder session record is created whenever its first child assessment shows up. But that record won't have its own adherence data updated until its own related session adherence record comes through.